implements `reindex` on models and Mongoid searches

tonyta opened the pull request:
Calling `reindex` on a Mongoid or MongoMapper model will reindex the
entire collection. When called on a Mongoid criteria, the reindex will
be scoped to that criteria.

Caveat: Calling `reindex` on a MongoMapper query will **NOT** scope the
reindex to the search. It will reindex the _entire_ collection.

dblock commented:
This seems reasonable. Lets start by fixing the build? You might want to do that as a separate PR as it looks like some dependencies have changed. I'll do a more thorough code review when it's green.

tonyta committed:
implements `reindex` on models and Mongoid searches

Calling reindex on a Mongoid or MongoMapper model will reindex the
entire collection. When called on a Mongoid criteria, the reindex will
be scoped to that criteria.

Caveat: Calling reindex on a MongoMapper query will NOT scope the
reindex to the search. It will reindex the entire collection.

tonyta commented:
Whoops! I got a little overzealous with `spec_helper`. The commit should be clean now.

Unfortunately, it looks like Bundler is having a hard time resolving dependencies that have dropped support for earlier versions of Ruby (`listen` and `mime-types`). I went ahead and added these in separate commits to allow the tests to run.

Would you consider any of the following?:
- Dropping support for Ruby < v2.1.0. (Ruby 2.0.0 has reached EoL this year.)
- Dropping sopport for Mongoid 3 (Rails 3 has effectively reached EoL)
- Dropping support for MongoMapper (both it and Mongoid have diverging implementations)

Thanks, @dblock!

tonyta commented:
I cannot locally reproduce the failing build with `MONGOID_VERSION=3.0` and `RUBY_VERSION=2.2`. Any suggestions? Would `master` still be green if rebuilt now?

dblock commented:
Definitely feel free to drop Ruby < 2.1.

I am open to dropping support for Mongoid 3 if it's too hard to maintain, but I wouldn't go out of the way to do this.

I'd rather not drop mongomapper support, if we think it's a good idea I would fork this gem into `subspot_mongoid` and continue as that. Feel free to do it, it might be a good idea.

dblock commented:
Add a line to CHANGELOG please. Might also need a README update.

dblock commented:
I also cannot reproduce the `require 'rails'` problem, but I would try to change it to `require 'rails/all'` and see if it makes any difference.

dblock commented:
I've opened https://github.com/derekharmel/sunspot_mongo/pull/15 to experiment a bit.

dblock commented:
I merged #15 as it went green, you can rebase.

tonyta commented:
Okay! Rebased and ready to merge!

As for `MongoMapper`, you're right. It's probably not too hard to keep it. If this gem gets more complicated, breaking out separate adapters might help mitigate the ternaries and branching.

Thanks for the quick response and for maintaining this gem!

tonyta commented:
whoops... just saw your comment about the CHANGELOG and README. standby...

dblock commented:
Merged.
